### PR TITLE
Email Report - PDF First Draft!

### PR DESCRIPTION
**Pull request by rhKashif**

### CODE REVIEW

File: report/main.py [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+from psycopg2 import connect
+from psycopg2.extensions import connection
+
+
+from email.mime.multipart import MIMEMultipart
+from email.mime.application import MIMEApplication
+
+from xhtml2pdf import pisa
+import plotly_express as px
+import boto3
```
Comment (angelikavilde) [inline report/main.py]: from my understanding, these should not be separated with newline

File: report/main.py [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+    Returns:
+        DataFrame:  A pandas DataFrame containing all relevant release data
+    """
+    query = f""
+    df_releases = pd.read_sql_query(query, conn_postgres)
+
+    return df_releases
+
+
+def convert_html_to_pdf(source_html, output_filename):
```
Comment (angelikavilde) [inline report/main.py]: typehints missing

File: report/main.py [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+        str: A string relating to the title of the highest rated new game released
+    """
+    date = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")
+    df_releases = df_releases[df_releases["release_date"] == date]
+    df_ratings = df_releases.groupby("title")["sentiment"].mean().reset_index()
+
+    return df_ratings.head(1)["title"][0]
+
+
+def trending_game_information(df_releases: DataFrame, index: int) -> str:
```
Comment (angelikavilde) [inline report/main.py]: quite a bulky function, can you separate things out?

File: report/main.py [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+        - Windows: {windows_compatibility}<br>
+        - Linux: {linux_compatibility}<br>
+    </body>
+    </html>
+    """
+
+    return html_template
+
+
+def create_report(df_releases: DataFrame):
```
Comment (angelikavilde) [inline report/main.py]: returns None + can be separated into multiple functions

File: report/main.py [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+    </html>
+    '''
+
+    with open("test.html", "w") as file:
+        file.write(template)
+
+    convert_html_to_pdf(template, environ.get("REPORT_FILE"))
+
+
+def send_email():
```
Comment (angelikavilde) [inline report/main.py:356]: I'm assuming typehints and docstrings aren't finished so ignore some of the above

File: report/requirements.txt [hunk 6]
```diff
@@ -0,0 +1,9 @@
+altair
+python-dotenv
+psycopg2
```
Comment (angelikavilde) [inline report/requirements.txt]: -binary

File: report/main.py (hunk 1 above)
Comment (rhKashif) [reply report/main.py]: Oh yes, in my html induced depression I forgot to format this! Will normalise.

**Review by DMcCallion — APPROVED** [review]

File: dashboard/requirements.txt [hunk 7]
```diff
@@ -1,4 +1,6 @@
 altair
```
Comment (gawsalya) [inline dashboard/requirements.txt:1]: so minor but can this be alphabetical

File: report/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,9 @@
+altair
```
Comment (gawsalya) [inline report/requirements.txt:1]: can this be alphabetical please!

File: report/main.py [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+    return df_releases
+
+
+def convert_html_to_pdf(source_html, output_filename):
+    """
+    """
+    # open output file for writing (truncated binary)
+    result_file = open(output_filename, "w+b")
+
+    # convert HTML to PDF
```
Comment (gawsalya) [inline report/main.py]: what a weird library

File: report/main.py [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+        source_html,                # the HTML to convert
+        dest=result_file)           # file handle to recieve result
+
+    # close output file
+    result_file.close()                 # close output file
+
+    return pisa_status.err
+
+
+def get_new_releases(df_releases: DataFrame) -> int:
```
Comment (gawsalya) [inline report/main.py]: maybe get_number_of_new_releases it more appropriate

File: report/main.py [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+
+    Args: 
+        df_releases (DataFrame): A pandas DataFrame containing all relevant game data
+
+        index (int): A int associated with a number index within the trending game list
+
+    Returns:
+        str: A string relating to the information of selected trending new game released in html format
+    """
+    date = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")
```
Comment (gawsalya) [inline report/main.py]: 123 and 124 can be a new function as you've done it in the function before

File: report/main.py [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+    """
+    Return the name of new release with the highest overall sentiment score for the previous day
+
+    Args: 
+        df_releases (DataFrame): A pandas DataFrame containing all relevant game data
+
+    Returns:
+        str: A string relating to the title of the highest rated new game released
+    """
+    date = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")
```
Comment (gawsalya) [inline report/main.py]: 104 and 105 -- as mentioned below make into a new function as you've got it repeated

File: report/main.py [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+        str: A string relating to the information of selected trending new game released in html format
+    """
+    date = (datetime.now() - timedelta(days=1)).strftime("%Y-%m-%d")
+    df_releases = df_releases[df_releases["release_date"] == date]
+    df_trending_game = df_releases.groupby(
+        "title")["sentiment"].mean().reset_index().iloc[index]
+    release_name = df_trending_game["title"]
+    sentiment = round(df_trending_game["sentiment"], 1)
+    release_df = df_releases[df_releases["title"]
+                             == f"{release_name}"].reset_index().head(1)
```
Comment (gawsalya) [inline report/main.py:147]: Curiosity Q: Oo what is this single equal sign then double equal sign!

File: report/main.py (hunk 3 above)
Comment (gawsalya) [reply report/main.py]: do you see anywhere else than can be separated out -- I can only see one bit really but think the rest is relevant to making the html hm hm

File: report/main.py [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+
+def send_email():
+    """
+    """
+    client = boto3.client("ses",
+                          region_name="eu-west-2",
+                          aws_access_key_id=environ["AWS_ACCESS_KEY_ID"],
+                          aws_secret_access_key=environ["AWS_SECRET_ACCESS_KEY"])
+
+    message = MIMEMultipart()
```
Comment (gawsalya) [inline report/main.py:371]: another strange library, seems you have learnt a lot of new things! Good job!

**Review by gawsalya — CHANGES_REQUESTED** [review]
Just made a few comments, looks good!

File: report/main.py (hunk 13 above)
Comment (angelikavilde) [reply report/main.py:147]: @gawsalya , in here he sets release_df to be a part of df_releases where the title is specifically the release_name variable

File: report/main.py (hunk 3 above)
Comment (rhKashif) [reply report/main.py]: Separated as much as possible for now but will have to revisit when integrating live data in next draft.

File: report/main.py (hunk 14 above)
Comment (rhKashif) [reply report/main.py:371]: All ugly libraries!

**Pull request merged**